### PR TITLE
KEP: Add an exponential backoff mechanism to the requeueing strategy

### PR DESCRIPTION
**Pull request by tenzen-y**
#### What type of PR is this?

/kind documentation

#### Which issue(s) this PR fixes:

Part-of: #1282

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: keps/1282-pods-ready-requeue-strategy/README.md [hunk 1]
```diff
@@ -56,6 +56,22 @@ Consider the following scenario:
 
 In this case, the administrator would like the evicted workload to be requeued as soon as possible on the newly available capacity.
 
+#### Story 2
+
+In the story 1 scenario, when we set `waitForPodsReady.requeuingTimestamp=Eviction`,
```
Comment (nstogner) [inline keps/1282-pods-ready-requeue-strategy/README.md:61]: I am guessing this should be `waitForPodsReady.requeuingTimestamp=Creation` ... which would preserve the pre-eviction ordering (i.e. put the evicted workload at the front of the queue again).

File: keps/1282-pods-ready-requeue-strategy/README.md [hunk 2]
```diff
@@ -56,6 +56,22 @@ Consider the following scenario:
 
 In this case, the administrator would like the evicted workload to be requeued as soon as possible on the newly available capacity.
 
+#### Story 2
+
+In the story 1 scenario, when we set `waitForPodsReady.requeuingTimestamp=Eviction`, 
+the workload endlessly or repeatedly can be put in front of the queue after eviction in the following eviction reasons:
+
+1. The workload don't have the proper configurations like image pull credential and pvc name, etc.
+2. The cluster can meet flavorQuotas, but each node doesn't have the resources that each podSet requests.
```
Comment (nstogner) [inline keps/1282-pods-ready-requeue-strategy/README.md:65]: 3. If there are multiple resource flavors that match the workload (flavors 1 & 2) and the workload is running on flavor 2 (where there happens to be a quota issue - i.e. flavor 2 is backed by on-demand capacity): there is a high likelihood that the workload will be requeued on the same flavor - with the same capacity issue (because other workloads will have taken up resource flavor 1 while the requeued workload was running.

File: keps/1282-pods-ready-requeue-strategy/README.md (hunk 1 above)
Comment (tenzen-y) [reply keps/1282-pods-ready-requeue-strategy/README.md:61]: Oops, it's a great catch! Thank you!

File: keps/1282-pods-ready-requeue-strategy/README.md (hunk 2 above)
Comment (tenzen-y) [reply keps/1282-pods-ready-requeue-strategy/README.md:65]: @nstogner Sorry, I couldn't understand when eviction happens, who is the preemptor, and who is the evictor.

Could you clarify?

File: keps/1282-pods-ready-requeue-strategy/README.md (hunk 1 above)
Comment (tenzen-y) [reply keps/1282-pods-ready-requeue-strategy/README.md]: Done.

File: keps/1282-pods-ready-requeue-strategy/README.md [hunk 3]
```diff
@@ -75,25 +105,69 @@ Consider including folks who also work outside the SIG or subproject.
...
-    kind: Configuration
-    # ...
-    waitForPodsReady:
-      requeuingTimestamp: Creation | Eviction # <-- New field
+```go
+type WaitForPodsReady struct {
+	...
+	// requeuingStrategy defines the strategy for requeuing a Workload
+	// +optional
+	RequeuingStrategy *RequeuingStrategy `json:"requeuingStrategy,omitempty"`
```
Comment (tenzen-y) [inline keps/1282-pods-ready-requeue-strategy/README.md:122]: I think that we can break this API for now since we haven't yet released this API.

**Comment by tenzen-y** [comment]
/hold
for reviews.

**Comment by tenzen-y** [comment]
Due to the failure to set up the webhook server: 

> A BeforeSuite node failed so all tests were skipped.

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_kueue/1608/pull-kueue-test-e2e-main-1-29/1751021175854600192

/test pull-kueue-test-e2e-main-1-29

File: keps/1282-pods-ready-requeue-strategy/README.md (hunk 2 above)
Comment (alculquicondor) [reply keps/1282-pods-ready-requeue-strategy/README.md:73]: @nstogner case is still about waitForPodsReady timeout. Once the workload is evicted because it didn't become ready, there is the possibility that it won't fit in the first flavor, so it will be assigned back to the second flavor.

File: keps/1282-pods-ready-requeue-strategy/README.md [hunk 4]
```diff
@@ -56,6 +64,28 @@ Consider the following scenario:
...
 
+#### Story 2
+
+In the story 1 scenario, when we set `waitForPodsReady.requeuingStrategy.timestamp=Creation`, 
+the workload endlessly or repeatedly can be put in front of the queue after eviction in the following eviction reasons:
+
+1. The workload don't have the proper configurations like image pull credential and pvc name, etc.
+2. The cluster can meet flavorQuotas, but each node doesn't have the resources that each podSet requests.  
+
+Specifically, the second reason will often occur in the workloads required gpus.
```
Comment (alculquicondor) [inline keps/1282-pods-ready-requeue-strategy/README.md:75]:
```suggestion
Specifically, the second reason will often occur if the available quota is fragmented across multiple nodes,
such that the workload can't be scheduled in a node even though there is enough
quota in the cluster.
For example,
```

File: keps/1282-pods-ready-requeue-strategy/README.md [hunk 5]
```diff
@@ -56,6 +64,28 @@ Consider the following scenario:
...
+has 2 worker nodes with 4 gpus, and 3 gpus are used (which means 1 gpu is free in each node),
+the workload will be repeatedly evicted because of the lack of resources in each node even though the cluster has enough capacities.
+
+In this case, to avoid rapid repetition of the admission and eviction cycle,
+the administrator would like to use an exponential backoff mechanism and add a maximum number of retries.
+
+#### Story 3
+
+In the story 3 scenario, after the evicted workload reached the maximum retry condition, 
+we want to easily requeue the workload to the queue without recreating the job.
```
Comment (alculquicondor) [inline keps/1282-pods-ready-requeue-strategy/README.md:86]: And what is the high level solution for this?

File: keps/1282-pods-ready-requeue-strategy/README.md [hunk 6]
```diff
@@ -75,25 +105,69 @@ Consider including folks who also work outside the SIG or subproject.
...
+}
+
+type RequeuingStrategy struct {
+	// timestamp defines the timestamp used for requeuing a Workload
+	// that was evicted due to Pod readiness. Defaults to Eviction.
+	// +optional
+	Timestamp *RequeuingTimestamp `json:"timestamp,omitempty"`
+	
+	// backOffLimitCount defines the maximum number of requeuing retries.
+	// When the number is reached, the workload is deactivated.
```
Comment (alculquicondor) [inline keps/1282-pods-ready-requeue-strategy/README.md:132]: What do you mean by deactivated?

File: keps/1282-pods-ready-requeue-strategy/README.md [hunk 7]
```diff
@@ -75,25 +105,69 @@ Consider including folks who also work outside the SIG or subproject.
...
+	// +optional
+	BackOffLimitCount *int32 `json:"backOffLimitCount,omitempty"`
+    
+	// backOffLimitTimeout defines the time for a workload that 
+	// has once been admitted to reach the PodsReady=true condition. 
+	// When the time is reached, the workload is deactivated.
+	// 	
+	// Defaults to null.
+	// +optional
+	BackOffLimitTimeout *int32 `json:"backOffLimitTimeout,omitempty"`
```
Comment (alculquicondor) [inline keps/1282-pods-ready-requeue-strategy/README.md:144]: I don't understand what this is for. Can you rephrase?

File: keps/1282-pods-ready-requeue-strategy/README.md [hunk 8]
```diff
@@ -75,25 +105,69 @@ Consider including folks who also work outside the SIG or subproject.
...
+```
+
+#### Workload
+
+Add a new field, "requeuedCount", to the Workload to allow recording the number of times a workload is requeued.
+
+```go
+type WorkloadStatus struct {
+	...
+	// requeuedCount determines the number of times a workload has been requeued.
```
Comment (alculquicondor) [inline keps/1282-pods-ready-requeue-strategy/README.md:165]:
```suggestion
	// requeuedCount records the number of times a workload has been requeued.
```

File: keps/1282-pods-ready-requeue-strategy/README.md [hunk 9]
```diff
@@ -104,10 +178,40 @@ Currently, workloads within a ClusterQueue are sorted based on 1. Priority and 2
...
+The `pkg/workload` package could be modified to include a conditional (`if evictionReason == kueue.WorkloadEvictedByPodsReadyTimeout`) 
+that controls which timestamp to return based on the configured ordering strategy.
+The same sorting logic would also be used when sorting the heads of queues.
 
 Update the `apis/config/<version>` package to include `Creation` and `Eviction` constants.
 
+### Exponential Backoff Mechanism
+
+When the kueueConfig `backOffLimitCount` or `backOffLimitTimeout` is set and there are evicted workloads by waitForPodsReady,
+the queueManager returns evicted workloads that an exponential backoff duration finished and other workloads as a headWorkloads.
```
Comment (alculquicondor) [inline keps/1282-pods-ready-requeue-strategy/README.md:190]:
```suggestion
the queueManager holds evicted workloads with an exponential backoff. During this time, other workloads will have a chance to be admitted.
```

File: keps/1282-pods-ready-requeue-strategy/README.md [hunk 10]
```diff
@@ -104,10 +178,40 @@ Currently, workloads within a ClusterQueue are sorted based on 1. Priority and 2
...
+
+#### backOffLimitCount
+
+When a workload eviction is issued with `PodsReadyTimeout` condition, 
+a workload `.status.requeuedCount` is incremented by 1 each time　in the workload controller.
+
+After that, when a workload `.status.requeudCount` reaches the kueueConfig `.waitForPodsReady.requeueingStrategy.backOffLimitCount`,
+a workload is deactivated by setting false to `.spec.active` instead of be suspended in the jobframework reconciler.
+
+#### backOffLimitTimeout
```
Comment (alculquicondor) [inline keps/1282-pods-ready-requeue-strategy/README.md:204]: What is the use case here?

Why isn't the backoffLimitCount enough?

File: keps/1282-pods-ready-requeue-strategy/README.md (hunk 6 above)
Comment (tenzen-y) [reply keps/1282-pods-ready-requeue-strategy/README.md:132]: I  meant using "Suspend a running Job without requeueing" feature.

File: keps/1282-pods-ready-requeue-strategy/README.md (hunk 2 above)
Comment (nstogner) [reply keps/1282-pods-ready-requeue-strategy/README.md:73]: Yes, my comment was in regards to the situation where eviction was due to waitForPodsReady. If the requeuing timestamp is configured to be the creation timestamp, and all resource flavors are in use, without backoff there is a very high likelihood that the freshly evicted workload will be readmitted into the same resource flavor, resulting in flapping.

File: keps/1282-pods-ready-requeue-strategy/README.md (hunk 10 above)
Comment (tenzen-y) [reply keps/1282-pods-ready-requeue-strategy/README.md:204]: > Why isn't the backoffLimitCount enough?

Because `backoffCount` is difficult to estimate how many hours jobs will actually be retrying. So, to improve the possibility for estimation, I added timeout separate from count.

If you think that we shouldn't have `backoffLimitTimeout` separate from `backoffLimitCount`, I'm ok with removing `backoffLimitTimeout`.

@alculquicondor WDYT?

File: keps/1282-pods-ready-requeue-strategy/README.md (hunk 2 above)
Comment (tenzen-y) [reply keps/1282-pods-ready-requeue-strategy/README.md:73]: >  case is still about waitForPodsReady timeout. Once the workload is evicted because it didn't become ready, there is the possibility that it won't fit in the first flavor, so it will be assigned back to the second flavor.

Oh, it makes sense. I will add this story here.
@alculquicondor @nstogner Thank you.

File: keps/1282-pods-ready-requeue-strategy/README.md (hunk 10 above)
Comment (tenzen-y) [reply keps/1282-pods-ready-requeue-strategy/README.md:204]: SGTM
We can start only support `backoffCount` for now.

File: keps/1282-pods-ready-requeue-strategy/README.md (hunk 7 above)
Comment (tenzen-y) [reply keps/1282-pods-ready-requeue-strategy/README.md:144]: I will remove this knob as we discussed here: https://github.com/kubernetes-sigs/kueue/pull/1608#discussion_r1475161093

File: keps/1282-pods-ready-requeue-strategy/README.md (hunk 4 above)
Comment (tenzen-y) [reply keps/1282-pods-ready-requeue-strategy/README.md]: Done.

File: keps/1282-pods-ready-requeue-strategy/README.md (hunk 8 above)
Comment (tenzen-y) [reply keps/1282-pods-ready-requeue-strategy/README.md]: Done.

File: keps/1282-pods-ready-requeue-strategy/README.md (hunk 9 above)
Comment (tenzen-y) [reply keps/1282-pods-ready-requeue-strategy/README.md]: Done.

File: keps/1282-pods-ready-requeue-strategy/README.md (hunk 5 above)
Comment (alculquicondor) [reply keps/1282-pods-ready-requeue-strategy/README.md:88]: Maybe put in the story something like:

"This is possible if the Workload is deactivated (active=false), as opposed to deleting it".

File: keps/1282-pods-ready-requeue-strategy/README.md (hunk 6 above)
Comment (alculquicondor) [reply keps/1282-pods-ready-requeue-strategy/README.md:134]:
```suggestion
	// When the number is reached, the workload is deactivated (`.spec.active=false`).	
```

File: keps/1282-pods-ready-requeue-strategy/README.md (hunk 7 above)
Comment (alculquicondor) [reply keps/1282-pods-ready-requeue-strategy/README.md:146]: Please remove so we can move forward with the KEP.

File: keps/1282-pods-ready-requeue-strategy/README.md [hunk 11]
```diff
@@ -75,25 +107,69 @@ Consider including folks who also work outside the SIG or subproject.
...
+	// that was evicted due to Pod readiness. Defaults to Eviction.
+	// +optional
+	Timestamp *RequeuingTimestamp `json:"timestamp,omitempty"`
+	
+	// backOffLimitCount defines the maximum number of requeuing retries.
+	// When the number is reached, the workload is deactivated.	
+	//
+	// Defaults to null. 
+	// +optional
+	BackOffLimitCount *int32 `json:"backOffLimitCount,omitempty"`
```
Comment (alculquicondor) [inline keps/1282-pods-ready-requeue-strategy/README.md:138]: Please describe the implicit behavior of the backoff. What is the base backoff, and what is the exponent?

File: keps/1282-pods-ready-requeue-strategy/README.md [hunk 12]
```diff
@@ -75,25 +107,69 @@ Consider including folks who also work outside the SIG or subproject.
...
+Add a new field, "requeuedCount", to the Workload to allow recording the number of times a workload is requeued.
+
+```go
+type WorkloadStatus struct {
+	...
+	// requeuedCount records the number of times a workload has been requeued.
+	// When a deactivated workload is reactivated, this count is reset to 0. 
+	//
+	// +optional
+	RequeuedCount *int32 `json:"requeuedCount,omitempty"`
```
Comment (astefanutti) [inline keps/1282-pods-ready-requeue-strategy/README.md:171]: How about adding the last requeue / backoff timestamp and the backoff duration, so it's possible to know when the workload will be "retried", without recomputing the exponential backoff logic for an external observer?

Comment (astefanutti) [inline keps/1282-pods-ready-requeue-strategy/README.md:171]: What are the invariants that make a workload in "backoff" state? Is it a combination of `PodsReadyTimeout` condition and `.status.requeuedCount > 0`?

Comment (alculquicondor) [reply keps/1282-pods-ready-requeue-strategy/README.md:171]: +1
I wanted to keep the API as minimal as possible, but I agree that usability can be compromised without a `maxBackoff`

File: keps/1282-pods-ready-requeue-strategy/README.md (hunk 5 above)
Comment (tenzen-y) [reply keps/1282-pods-ready-requeue-strategy/README.md:88]: SGTM

File: keps/1282-pods-ready-requeue-strategy/README.md (hunk 11 above)
Comment (tenzen-y) [reply keps/1282-pods-ready-requeue-strategy/README.md:138]: It makes sense.

File: keps/1282-pods-ready-requeue-strategy/README.md (hunk 12 above)
Comment (tenzen-y) [reply keps/1282-pods-ready-requeue-strategy/README.md:171]: SGTM
I will extend this API.

Comment (tenzen-y) [reply keps/1282-pods-ready-requeue-strategy/README.md:171]: Yes, that's right. Those parameters aren't configurable.

File: keps/1282-pods-ready-requeue-strategy/README.md (hunk 7 above)
Comment (tenzen-y) [reply keps/1282-pods-ready-requeue-strategy/README.md]: Done.

File: keps/1282-pods-ready-requeue-strategy/README.md (hunk 10 above)
Comment (tenzen-y) [reply keps/1282-pods-ready-requeue-strategy/README.md]: Done.

File: keps/1282-pods-ready-requeue-strategy/README.md (hunk 2 above)
Comment (tenzen-y) [reply keps/1282-pods-ready-requeue-strategy/README.md:76]: Done.

File: keps/1282-pods-ready-requeue-strategy/README.md (hunk 6 above)
Comment (tenzen-y) [reply keps/1282-pods-ready-requeue-strategy/README.md:138]: Done.

File: keps/1282-pods-ready-requeue-strategy/README.md (hunk 5 above)
Comment (tenzen-y) [reply keps/1282-pods-ready-requeue-strategy/README.md:92]: Done.

File: keps/1282-pods-ready-requeue-strategy/README.md (hunk 12 above)
Comment (tenzen-y) [reply keps/1282-pods-ready-requeue-strategy/README.md:166]: Ah, there is one more thing. We won't expose the parameters to initialize Backoff: https://pkg.go.dev/k8s.io/apimachinery/pkg/util/wait@v0.29.1#Backoff

File: keps/1282-pods-ready-requeue-strategy/README.md [hunk 13]
```diff
@@ -56,6 +66,31 @@ Consider the following scenario:
...
 In this case, the administrator would like the evicted workload to be requeued as soon as possible on the newly available capacity.
 
+#### Story 2
+
+In the story 1 scenario, when we set `waitForPodsReady.requeuingStrategy.timestamp=Creation`, 
+the workload endlessly or repeatedly can be put in front of the queue after eviction in the following eviction reasons:
+
+1. The workload don't have the proper configurations like image pull credential and pvc name, etc.
+2. The cluster can meet flavorQuotas, but each node doesn't have the resources that each podSet requests.  
+3. Multiple flavors are matched for the workload, but the workload can not be launched on the backed flavors (which means non-primary flavor).
```
Comment (alculquicondor) [inline keps/1282-pods-ready-requeue-strategy/README.md:76]: was this always there? I don't really understand it.

File: keps/1282-pods-ready-requeue-strategy/README.md (hunk 12 above)
Comment (alculquicondor) [reply keps/1282-pods-ready-requeue-strategy/README.md:166]: can you add it? We are almost closing the release.

Comment (tenzen-y) [reply keps/1282-pods-ready-requeue-strategy/README.md:166]: I should do ASAP...

File: keps/1282-pods-ready-requeue-strategy/README.md (hunk 13 above)
Comment (tenzen-y) [reply keps/1282-pods-ready-requeue-strategy/README.md:76]: I added this story built with https://github.com/kubernetes-sigs/kueue/pull/1608/files#r1460075048.

File: keps/1282-pods-ready-requeue-strategy/README.md (hunk 12 above)
Comment (tenzen-y) [reply keps/1282-pods-ready-requeue-strategy/README.md:166]: @alculquicondor Here are quick API design sharing: Please let me know if the API design wouldn't be expected.

```golang
type WorkloadStatus struct {
	...
	// requeueState records 
	// 
	// +optional
	RequeueState *RequeueState `json:"requeueState,omitempty"` 
}

type RequeueState struct {
	// count records the number of times a workload has been requeued.
	// When a deactivated workload is reactivated, this count is reset to 0. 
	//
	// +optional
	Count *int32 `json:"requeuedCount,omitempty"`

	// +optional
	LastBackoffTime metav1.Time `json:"lastRequeuedTime,omitempty"`
	
	// +optional
	BackoffDuration metav1.Duration `json:"backoffDuration,omitempty"`
}
```

I'm organizing the above API.

Comment (alculquicondor) [reply keps/1282-pods-ready-requeue-strategy/README.md:166]: instead of `BackoffDuration` and `lastBackoffTime`, it might be better to have just `requeueAt metav1.Time` that indicates the time at which Kueue will consider this workload again for admission.

The time when the workload was evicted is already visible in the condition.

Comment (tenzen-y) [reply keps/1282-pods-ready-requeue-strategy/README.md:166]: > instead of BackoffDuration and lastBackoffTime, it might be better to have just requeueAt metav1.Time that indicates the time at which Kueue will consider this workload again for admission.
> 
> The time when the workload was evicted is already visible in the condition.

That makes sense. Even if we add only `requeueAt`, we can avoid recomputing the exponential backoff logic.
After we truly want to have `duration` based on user feedback, we can extend API.

File: keps/1282-pods-ready-requeue-strategy/README.md (hunk 13 above)
Comment (alculquicondor) [reply keps/1282-pods-ready-requeue-strategy/README.md:76]: oh right. Let me rephrase:

```
If there are multiple resource flavors that match the workload (for example, flavors 1 & 2) and the workload was running on flavor 2, it's likely that the workload will be readmitted on the same flavor.
```

Comment (tenzen-y) [reply keps/1282-pods-ready-requeue-strategy/README.md:76]: Thank you for this suggestion!

File: keps/1282-pods-ready-requeue-strategy/README.md (hunk 12 above)
Comment (tenzen-y) [reply keps/1282-pods-ready-requeue-strategy/README.md]: Done.

File: keps/1282-pods-ready-requeue-strategy/README.md (hunk 13 above)
Comment (tenzen-y) [reply keps/1282-pods-ready-requeue-strategy/README.md]: Done.

**Comment by tenzen-y** [comment]
Pending is only here: https://github.com/kubernetes-sigs/kueue/pull/1608#discussion_r1478419459
